File: virtual_computer_c/neuromorphic_simulator.c
/*
 * Neuromorphic Processor Simulator - C Implementation
 */

#include "neuromorphic_simulator.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <math.h>
#include <time.h>

#define INITIAL_NEURONS_CAPACITY 1024
#define INITIAL_SYNAPSES_CAPACITY 8192
#define INITIAL_SPIKE_LOG_CAPACITY 4096

/**
 * Create and initialize neuromorphic processor
 */
neuromorphic_processor_t* nm_create(uint32_t num_neurons, uint32_t num_layers) {
    neuromorphic_processor_t *nm = (neuromorphic_processor_t *)malloc(sizeof(neuromorphic_processor_t));
    if (!nm) return NULL;
    
    nm->num_neurons = num_neurons;
    nm->num_layers = num_layers;
    
    /* Allocate neurons */
    nm->neurons = (neuron_t *)malloc(num_neurons * sizeof(neuron_t));
    if (!nm->neurons) {
        free(nm);
        return NULL;
    }
    nm->neurons_capacity = num_neurons;
    
    /* Initialize neurons */
    uint32_t neurons_per_layer = num_neurons / num_layers;
    srand(time(NULL));
    
    for (uint32_t i = 0; i < num_neurons; i++) {
        neuron_t *neuron = &nm->neurons[i];
        neuron->neuron_id = i;
        neuron->neuron_type = NEURON_LIF;
        neuron->layer = i / neurons_per_layer;
        
        /* LIF parameters */
        neuron->membrane_potential = 0.0;
        neuron->threshold = 1.0 + (rand() / (double)RAND_MAX - 0.5) * 0.4;
        neuron->resting_potential = -70.0;
        neuron->tau_membrane = 20.0;
        neuron->tau_synaptic = 5.0;
        
        /* Tracking */
        neuron->spike_count = 0;
        neuron->last_spike_time = 0.0;
        neuron->synaptic_input = 0.0;
        neuron->refractory_period = 2.0;
        neuron->in_refractory = false;
    }
    
    /* Allocate synapses */
    nm->synapses = (synapse_t *)malloc(INITIAL_SYNAPSES_CAPACITY * sizeof(synapse_t));
    if (!nm->synapses) {
        free(nm->neurons);
        free(nm);
        return NULL;
    }
    nm->num_synapses = 0;
    nm->synapses_capacity = INITIAL_SYNAPSES_CAPACITY;
    
    /* Allocate spike log */
    nm->spike_log = (spike_event_t *)malloc(INITIAL_SPIKE_LOG_CAPACITY * sizeof(spike_event_t));
    if (!nm->spike_log) {
        free(nm->synapses);
        free(nm->neurons);
        free(nm);
        return NULL;
    }
    nm->spike_log_size = 0;
    nm->spike_log_capacity = INITIAL_SPIKE_LOG_CAPACITY;
    
    /* Time tracking */
    nm->current_time = 0.0;
    nm->time_step = 1.0;
    
    /* Statistics */
    nm->total_spikes = 0;
    nm->total_simulation_steps = 0;
    nm->energy_consumed_uj = 0.0;
    nm->latency_ms = 0.0;
    
    /* Create connectivity */
    nm_create_connectivity(nm, 0.1);
    
    return nm;
}

/**
 * Destroy neuromorphic processor
 */
void nm_destroy(neuromorphic_processor_t *nm) {
    if (!nm) return;
    free(nm->neurons);
    free(nm->synapses);
    free(nm->spike_log);
    free(nm);
}

/**
 * Create random connectivity between neurons
 */
void nm_create_connectivity(neuromorphic_processor_t *nm, double connectivity_ratio) {
    if (!nm) return;
    
    uint32_t target_connections = (uint32_t)(nm->num_neurons * connectivity_ratio);
    
    for (uint32_t i = 0; i < target_connections; i++) {
        uint32_t pre = rand() % nm->num_neurons;
        uint32_t post = rand() % nm->num_neurons;
        
        if (pre != post) {
            /* Check for duplicate */
            bool exists = false;
            for (uint32_t j = 0; j < nm->num_synapses; j++) {
                if (nm->synapses[j].pre_neuron_id == pre && nm->synapses[j].post_neuron_id == post) {
                    exists = true;
                    break;
                }
            }
            
            if (!exists) {
                /* Resize if needed */
                if (nm->num_synapses >= nm->synapses_capacity) {
                    nm->synapses_capacity *= 2;
                    nm->synapses = (synapse_t *)realloc(nm->synapses,
                        nm->synapses_capacity * sizeof(synapse_t));
                    if (!nm->synapses) return;
                }
                
                synapse_t *syn = &nm->synapses[nm->num_synapses];
                syn->pre_neuron_id = pre;
                syn->post_neuron_id = post;
                syn->weight = 0.1 + (rand() / (double)RAND_MAX) * 0.9;
                syn->delay = 1.0;
                syn->learning_rate = 0.01;
                syn->last_update = 0.0;
                syn->pre_spike_trace = 0.0;
                syn->post_spike_trace = 0.0;
                syn->plasticity_enabled = true;
                
                nm->num_synapses++;
            }
        }
    }
}

/**
 * Inject spikes into neurons (input)
 */
void nm_inject_spikes(neuromorphic_processor_t *nm, const uint32_t *neuron_ids,
                     uint32_t count, double current_time) {
    if (!nm || !neuron_ids) return;
    
    for (uint32_t i = 0; i < count; i++) {
        uint32_t nid = neuron_ids[i];
        if (nid < nm->num_neurons) {
            nm->neurons[nid].synaptic_input += 2.0;
            
            /* Record spike */
            if (nm->spike_log_size >= nm->spike_log_capacity) {
                nm->spike_log_capacity *= 2;
                nm->spike_log = (spike_event_t *)realloc(nm->spike_log,
                    nm->spike_log_capacity * sizeof(spike_event_t));
                if (!nm->spike_log) return;
            }
            
            nm->spike_log[nm->spike_log_size].neuron_id = nid;
            nm->spike_log[nm->spike_log_size].timestamp = current_time;
            nm->spike_log[nm->spike_log_size].source_layer = 0;
            nm->spike_log_size++;
        }
    }
}

/**
 * Update neuron state using LIF model
 */
bool nm_update_neuron(neuromorphic_processor_t *nm, uint32_t neuron_id,
                     double current_time) {
    if (!nm || neuron_id >= nm->num_neurons) return false;
    
    neuron_t *neuron = &nm->neurons[neuron_id];
    
    /* Skip if in refractory period */
    if (neuron->in_refractory) {
        if (current_time - neuron->last_spike_time > neuron->refractory_period) {
            neuron->in_refractory = false;
        } else {
            return false;
        }
    }
    
    /* LIF dynamics */
    double decay = exp(-nm->time_step / neuron->tau_membrane);
    neuron->membrane_potential = neuron->resting_potential +
        (neuron->membrane_potential - neuron->resting_potential) * decay +
        neuron->synaptic_input * 10.0 * (1.0 - decay);
    
    /* Decay synaptic input */
    neuron->synaptic_input *= exp(-nm->time_step / neuron->tau_synaptic);
    
    /* Check threshold */
    bool did_spike = false;
    if (neuron->membrane_potential >= neuron->threshold) {
        neuron->spike_count++;
        neuron->last_spike_time = current_time;
        neuron->in_refractory = true;
        neuron->membrane_potential = neuron->resting_potential;
        
        /* Record spike */
        if (nm->spike_log_size >= nm->spike_log_capacity) {
            nm->spike_log_capacity *= 2;
            nm->spike_log = (spike_event_t *)realloc(nm->spike_log,
                nm->spike_log_capacity * sizeof(spike_event_t));
            if (!nm->spike_log) return false;
        }
        
        nm->spike_log[nm->spike_log_size].neuron_id = neuron_id;
        nm->spike_log[nm->spike_log_size].timestamp = current_time;
        nm->spike_log[nm->spike_log_size].source_layer = neuron->layer;
        nm->spike_log_size++;
        
        nm->total_spikes++;
        did_spike = true;
    }
    
    return did_spike;
}

/**
 * Propagate spikes through synapses
 */
void nm_propagate_spikes(neuromorphic_processor_t *nm, double current_time) {
    if (!nm) return;
    
    /* Find recent spikes */
    for (uint32_t i = 0; i < nm->spike_log_size; i++) {
        if (nm->spike_log[i].timestamp > current_time - 5.0) {
            uint32_t spike_neuron = nm->spike_log[i].neuron_id;
            
            /* Find outgoing synapses */
            for (uint32_t j = 0; j < nm->num_synapses; j++) {
                if (nm->synapses[j].pre_neuron_id == spike_neuron) {
                    if (current_time - nm->spike_log[i].timestamp >= nm->synapses[j].delay) {
                        uint32_t post_id = nm->synapses[j].post_neuron_id;
                        nm->neurons[post_id].synaptic_input += nm->synapses[j].weight;
                    }
                }
            }
        }
    }
}

/**
 * Simulate one time step
 */
void nm_simulate_step(neuromorphic_processor_t *nm, double current_time, bool inject_input) {
    if (!nm) return;
    
    /* Inject random input if requested */
    if (inject_input) {
        uint32_t num_input = nm->num_neurons / 100;
        if (num_input < 1) num_input = 1;
        
        uint32_t *input_neurons = (uint32_t *)malloc(num_input * sizeof(uint32_t));
        if (input_neurons) {
            for (uint32_t i = 0; i < num_input; i++) {
                input_neurons[i] = rand() % nm->num_neurons;
            }
            nm_inject_spikes(nm, input_neurons, num_input, current_time);
            free(input_neurons);
        }
    }
    
    /* Update all neurons */
    uint32_t spikes_this_step = 0;
    for (uint32_t i = 0; i < nm->num_neurons; i++) {
        if (nm_update_neuron(nm, i, current_time)) {
            spikes_this_step++;
        }
    }
    
    /* Propagate spikes */
    nm_propagate_spikes(nm, current_time);
    
    /* Energy calculation */
    nm->energy_consumed_uj += spikes_this_step * 0.001;
    
    nm->total_simulation_steps++;
}

/**
 * Get average spike rate for a layer
 */
double nm_get_layer_spike_rate(neuromorphic_processor_t *nm, uint32_t layer) {
    if (!nm) return 0.0;
    
    uint32_t count = 0;
    uint64_t total_spikes = 0;
    
    for (uint32_t i = 0; i < nm->num_neurons; i++) {
        if (nm->neurons[i].layer == layer) {
            total_spikes += nm->neurons[i].spike_count;
            count++;
        }
    }
    
    return count > 0 ? (double)total_spikes / count : 0.0;
}

/**
 * Get processor statistics
 */
void nm_get_stats(neuromorphic_processor_t *nm, char *buffer, size_t buffer_size) {
    if (!nm || !buffer) return;
    
    double avg_spike_rate = nm->total_simulation_steps > 0 ?
        (double)nm->total_spikes / nm->total_simulation_steps : 0.0;
    
    snprintf(buffer, buffer_size,
        "Neuromorphic Processor Statistics\n"
        "  Total Neurons: %u\n"
        "  Total Synapses: %u\n"
        "  Total Spikes: %lu\n"
        "  Simulation Steps: %lu\n"
        "  Avg Spike Rate: %.2f Hz\n"
        "  Energy Consumed: %.2f µJ\n",
        nm->num_neurons,
        nm->num_synapses,
        nm->total_spikes,
        nm->total_simulation_steps,
        avg_spike_rate,
        nm->energy_consumed_uj
    );
}

/**
 * Get connectivity statistics
 */
void nm_get_connectivity_stats(neuromorphic_processor_t *nm, char *buffer, size_t buffer_size) {
    if (!nm || !buffer) return;
    
    double connectivity_ratio = nm->num_synapses > 0 ?
        (double)nm->num_synapses / (nm->num_neurons * nm->num_neurons) : 0.0;
    
    double avg_weight = 0.0;
    double min_weight = 1.0;
    double max_weight = 0.0;
    
    if (nm->num_synapses > 0) {
        for (uint32_t i = 0; i < nm->num_synapses; i++) {
            avg_weight += nm->synapses[i].weight;
            if (nm->synapses[i].weight < min_weight) {
                min_weight = nm->synapses[i].weight;
            }
            if (nm->synapses[i].weight > max_weight) {
                max_weight = nm->synapses[i].weight;
            }
        }
        avg_weight /= nm->num_synapses;
    }
    
    snprintf(buffer, buffer_size,
        "Neuromorphic Connectivity\n"
        "  Total Synapses: %u\n"
        "  Connectivity Ratio: %.4f\n"
        "  Avg Weight: %.3f\n"
        "  Min Weight: %.3f\n"
        "  Max Weight: %.3f\n",
        nm->num_synapses,
        connectivity_ratio,
        avg_weight,
        min_weight,
        max_weight
    );
}

/**
 * Print processor status
 */
void nm_print_status(neuromorphic_processor_t *nm) {
    if (!nm) return;
    
    char stats_buf[512];
    char conn_buf[512];
    
    nm_get_stats(nm, stats_buf, sizeof(stats_buf));
    nm_get_connectivity_stats(nm, conn_buf, sizeof(conn_buf));
    
    printf("\n");
    printf("================================================================================\n");
    printf("  Neuromorphic Processor Status\n");
    printf("================================================================================\n");
    printf("%s\n", stats_buf);
    printf("%s\n", conn_buf);
    printf("================================================================================\n\n");
}

File: virtual_computer/neuromorphic_simulator.py
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] #!/usr/bin/env python3
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] """
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] Neuromorphic Processor Simulator
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] Simulates spiking neural networks (SNNs) with neuron dynamics, synaptic plasticity,
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] and event-based processing
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] """
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] 
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] 
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] 
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] 
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] 
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] 
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] class NeuronType(Enum):
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED]     LEAKY_INTEGRATE_AND_FIRE = "lif"
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED]     HODGKIN_HUXLEY = "hh"
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED]     IZHIKEVICH = "iz"
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] 
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] 
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] @dataclass
# [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] # [REVIEWED] class Spike:
    """Represents a spike event"""
    neuron_id: int
    timestamp: float
    source_layer: int
    
    def __hash__(self):
        return hash((self.neuron_id, self.timestamp))


@dataclass
class Neuron:
    """Neuromorphic neuron model"""
    neuron_id: int
    neuron_type: NeuronType
    layer: int
    
    # LIF parameters
    membrane_potential: float = 0.0
    threshold: float = 1.0
    resting_potential: float = -70.0  # mV
    tau_membrane: float = 20.0  # ms (time constant)
    tau_synaptic: float = 5.0  # ms
    
    # Tracking
    spike_count: int = 0
    last_spike_time: float = 0.0
    synaptic_input: float = 0.0
    refractory_period: float = 2.0  # ms
    in_refractory: bool = False
    
    def reset(self):
        """Reset neuron to resting state"""
        self.membrane_potential = self.resting_potential
        self.synaptic_input = 0.0
        self.in_refractory = False


@dataclass
class Synapse:
    """Neural synapse with plasticity"""
    pre_neuron_id: int
    post_neuron_id: int
    weight: float = 0.5
    delay: float = 1.0  # ms
    learning_rate: float = 0.01
    
    # Plasticity tracking
    last_update: float = 0.0
    pre_spike_trace: float = 0.0
    post_spike_trace: float = 0.0
    plasticity_enabled: bool = True


class NeuromorphicProcessor:
    """Simulates a neuromorphic processor with spiking neural networks"""
    
    def __init__(self, num_neurons: int = 1000, num_layers: int = 4):
        self.num_neurons = num_neurons
        self.num_layers = num_layers
        self.neurons: Dict[int, Neuron] = {}
        self.synapses: Dict[Tuple[int, int], Synapse] = {}
        
        # Initialize neurons across layers
        neurons_per_layer = num_neurons // num_layers
        for layer in range(num_layers):
            for n in range(neurons_per_layer):
                neuron_id = layer * neurons_per_layer + n
                self.neurons[neuron_id] = Neuron(
                    neuron_id=neuron_id,
                    neuron_type=NeuronType.LEAKY_INTEGRATE_AND_FIRE,
                    layer=layer,
                    threshold=1.0 + random.uniform(-0.2, 0.2)
                )
        
        # Create random connectivity
        self.create_connectivity()
        
        # Spike tracking
        self.spike_log: List[Spike] = []
        self.current_time = 0.0
        self.time_step = 1.0  # ms
        
        # Statistics
        self.total_spikes = 0
        self.total_simulation_steps = 0
        self.energy_consumed_uj = 0.0  # microjoules
        self.latency_ms = 0.0
    
    def create_connectivity(self, connectivity_ratio: float = 0.1):
        """Create random synaptic connections"""
        neurons_list = list(self.neurons.keys())
        
        for _ in range(int(len(neurons_list) * connectivity_ratio)):
            pre_neuron = random.choice(neurons_list)
            post_neuron = random.choice(neurons_list)
            
            if pre_neuron != post_neuron:
                key = (pre_neuron, post_neuron)
                if key not in self.synapses:
                    self.synapses[key] = Synapse(
                        pre_neuron_id=pre_neuron,
                        post_neuron_id=post_neuron,
                        weight=random.uniform(0.1, 1.0)
                    )
    
    def inject_spikes(self, neuron_ids: List[int], current_time: float):
        """Inject spikes into neurons (input layer)"""
        for nid in neuron_ids:
            if nid in self.neurons:
                neuron = self.neurons[nid]
                neuron.synaptic_input += 2.0  # Input current
                
                # Record spike
                spike = Spike(neuron_id=nid, timestamp=current_time, source_layer=0)
                self.spike_log.append(spike)
    
    def update_neuron(self, neuron_id: int, current_time: float):
        """Update neuron state using LIF model"""
        neuron = self.neurons[neuron_id]
        
        # Skip if in refractory period
        if neuron.in_refractory:
            if current_time - neuron.last_spike_time > neuron.refractory_period:
                neuron.in_refractory = False
            else:
                return False
        
        # LIF dynamics: dV/dt = (V_rest - V + I*R) / tau_m
        decay = math.exp(-self.time_step / neuron.tau_membrane)
        neuron.membrane_potential = (
            neuron.resting_potential +
            (neuron.membrane_potential - neuron.resting_potential) * decay +
            neuron.synaptic_input * 10.0 * (1.0 - decay)
        )
        
        # Decay synaptic input
        neuron.synaptic_input *= math.exp(-self.time_step / neuron.tau_synaptic)
        
        # Check threshold
        did_spike = False
        if neuron.membrane_potential >= neuron.threshold:
            neuron.spike_count += 1
            neuron.last_spike_time = current_time
            neuron.in_refractory = True
            neuron.membrane_potential = neuron.resting_potential
            
            # Record spike
            spike = Spike(
                neuron_id=neuron_id,
                timestamp=current_time,
                source_layer=neuron.layer
            )
            self.spike_log.append(spike)
            self.total_spikes += 1
            did_spike = True
        
        return did_spike
    
    def propagate_spikes(self, current_time: float):
        """Propagate spikes through synapses"""
        # Find neurons that spiked in recent time window
        recent_spikes = [s for s in self.spike_log 
                        if s.timestamp > current_time - 5.0]
        
        for spike in recent_spikes:
            # Find outgoing synapses
            for (pre, post), synapse in self.synapses.items():
                if pre == spike.neuron_id:
                    # Apply synaptic delay
                    if current_time - spike.timestamp >= synapse.delay:
                        post_neuron = self.neurons[post]
                        post_neuron.synaptic_input += synapse.weight
                        
                        # Spike-timing-dependent plasticity (STDP)
                        if synapse.plasticity_enabled:
                            self.update_synapse_weight(synapse, current_time)
    
    def update_synapse_weight(self, synapse: Synapse, current_time: float):
        """Update synapse weight using STDP"""
        # Simplified STDP: Hebbian learning
        # If pre and post fire together, increase weight
        pre_neuron = self.neurons[synapse.pre_neuron_id]
        post_neuron = self.neurons[synapse.post_neuron_id]
        
        # Check if neurons recently spiked
        pre_recent = current_time - pre_neuron.last_spike_time < 10.0
        post_recent = current_time - post_neuron.last_spike_time < 10.0
        
        if pre_recent and post_recent:
            delta_w = synapse.learning_rate * 0.1
            synapse.weight = min(1.0, synapse.weight + delta_w)
        else:
            # Slight decay
            synapse.weight *= 0.99
    
    def simulate_step(self, current_time: float, inject_input: bool = False) -> Dict:
        """Simulate one time step"""
        self.current_time = current_time
        
        # Inject random input spikes (simulating sensory input)
        if inject_input:
            input_neurons = random.sample(
                [n for n in self.neurons.keys() if self.neurons[n].layer == 0],
                k=max(1, self.num_neurons // 100)
            )
            self.inject_spikes(input_neurons, current_time)
        
        # Update all neurons
        spikes_this_step = 0
        for neuron_id in self.neurons.keys():
            if self.update_neuron(neuron_id, current_time):
                spikes_this_step += 1
        
        # Propagate spikes
        self.propagate_spikes(current_time)
        
        # Energy calculation: roughly proportional to spike activity
        self.energy_consumed_uj += spikes_this_step * 0.001  # microjoules per spike
        
        self.total_simulation_steps += 1
        
        return {
            "time": current_time,
            "spikes_this_step": spikes_this_step,
            "total_spikes": self.total_spikes,
            "energy_uj": self.energy_consumed_uj,
        }
    
    def get_layer_spike_rate(self, layer: int) -> float:
        """Get average spike rate for a layer"""
        neurons_in_layer = [n for n in self.neurons.values() if n.layer == layer]
        if not neurons_in_layer:
            return 0.0
        
        avg_spikes = sum(n.spike_count for n in neurons_in_layer) / len(neurons_in_layer)
        return avg_spikes
    
    def get_stats(self) -> Dict:
        """Get processor statistics"""
        return {
            "total_neurons": self.num_neurons,
            "total_synapses": len(self.synapses),
            "total_spikes": self.total_spikes,
            "total_simulation_steps": self.total_simulation_steps,
            "energy_consumed_uj": self.energy_consumed_uj,
            "avg_spike_rate": self.total_spikes / max(1, self.total_simulation_steps),
            "spike_log_size": len(self.spike_log),
        }
    
    def get_connectivity_stats(self) -> Dict:
        """Get network connectivity statistics"""
        connectivity_ratio = len(self.synapses) / (self.num_neurons ** 2)
        avg_weight = sum(s.weight for s in self.synapses.values()) / max(1, len(self.synapses))
        
        return {
            "total_synapses": len(self.synapses),
            "connectivity_ratio": connectivity_ratio,
            "avg_synapse_weight": avg_weight,
            "max_synapse_weight": max((s.weight for s in self.synapses.values()), default=0),
            "min_synapse_weight": min((s.weight for s in self.synapses.values()), default=0),
        }
    
    def print_status(self):
        """Print processor status summary"""
        stats = self.get_stats()
        conn = self.get_connectivity_stats()
        
        print(f"\n{'='*70}")
        print(f"  Neuromorphic Processor Status")
        print(f"{'='*70}")
        print(f"  Neurons: {stats['total_neurons']}")
        print(f"  Synapses: {conn['total_synapses']} (connectivity: {conn['connectivity_ratio']:.2%})")
        print(f"  Total Spikes: {stats['total_spikes']}")
        print(f"  Avg Spike Rate: {stats['avg_spike_rate']:.2f} Hz")
        print(f"  Energy Consumed: {stats['energy_consumed_uj']:.2f} µJ")
        print(f"{'='*70}\n")
